services/src/main/java/edu/unc/lib/dl/cdr/services/rest/modify/ImportXMLController.java

```diff
@@ -40,51 +33,56 @@
 
 import edu.unc.lib.dl.acl.util.GroupsThreadStore;
-import edu.unc.lib.dl.update.BulkMetadataUpdateJob;
+import edu.unc.lib.dl.cdr.services.processing.XMLImportService;
 
 /**
- * Controller which accepts CDR bulk metadata update packages and begins update operations
- * 
+ * API endpoint that accepts an XML document containing objects' metadata, kicks off update operations for
+ * individual objects, and sends an email with a report of objects that were/were not able to be updated.
+ *
  * @author bbpennel
- * @date Jul 30, 2015
+ * @author harring
  */
 @Controller
 public class ImportXMLController {
     private static final Logger log = LoggerFactory.getLogger(ImportXMLController.class);
 
     @Autowired
-    private net.greghaines.jesque.client.Client jesqueClient;
-    @Autowired
-    private String bulkMetadataQueueName;
-    @Autowired
-    private String dataDir;
-    private Path storagePath;
-
-    @PostConstruct
-    public void init() throws IOException {
-        storagePath = Paths.get(dataDir + "/metadataImport/");
-        // Create the directory if it doesn't already exist
-        Files.createDirectories(storagePath);
-    }
+    private XMLImportService service;
 
-    @RequestMapping(value = "importXML", method = RequestMethod.POST)
-    public @ResponseBody Object importXML(@RequestParam("file") MultipartFile xmlFile,
-            HttpServletRequest request) throws Exception {
-        log.info("User {} has submitted a bulk metadata update package", GroupsThreadStore.getUsername());
-        Map<String, String> result = new HashMap<>();
+    /**
+     * Imports an XML document containing metadata for all objects specified by the request
+     *
+     * @param xmlFile
+     * @return response entity with result and response code
+     */
+    @RequestMapping(value = "/edit/importXML", method = RequestMethod.POST)
+    public @ResponseBody ResponseEntity<Object> importXML(@RequestParam("file") MultipartFile xmlFile) {
 
-        File importFile = File.createTempFile("import", ".xml", storagePath.toFile());
-        FileUtils.writeByteArrayToFile(importFile, xmlFile.getBytes());
+        String username = GroupsThreadStore.getUsername();
+        log.info("User {} has submitted a bulk metadata update package", username);
 
-        String emailAddress = GroupsThreadStore.getEmail();
+        Map<String, Object> result = new HashMap<>();
+        result.put("action", "import xml");
+        result.put("username", username);
 
-        Job job = new Job(BulkMetadataUpdateJob.class.getName(), null, emailAddress, GroupsThreadStore.getUsername(),
-                GroupsThreadStore.getGroups(), importFile.getAbsolutePath(), xmlFile.getOriginalFilename());
-        jesqueClient.enqueue(bulkMetadataQueueName, job);
+        String userEmail = GroupsThreadStore.getEmail();
+        try (InputStream importStream = xmlFile.getInputStream()) {
+            service.pushJobToQueue(result, importStream, username, userEmail);
```

Comment:
Hadn't noticed before now, but would you be willing to update this to pass in the AgentPrincipal versus having the service pull it from the current thread? That is how the other controllers are working, and it means the service doesn't have to be opinionated about how the groups are submitted. Also, the agent principal already contains the username so you could replace the username parameter with it.

Reply:
fixed